File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        int [] rows = new int[ matrix.length ] ,
               cols = new int[ matrix[0].length ] ;
        for ( int i = 0 ; i < matrix.length ; i++ ) {
            for ( int j = 0 ; j < matrix[0].length ; j++ ) {
                if ( matrix[i][j] != 0 ) continue ; 
                rows[i] = 1 ;
                cols[j] = 1 ;
            }
        }
        for ( int i = 0 ; i < matrix.length ; i++ ) {
            for ( int j = 0 ; j < matrix[0].length ; j++ ) {
                if ( rows[i] != 1 && cols[j] != 1 ) continue ; 
                matrix[i][j] = 0 ;
            }
        }
    }
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.kt
class Solution {
    fun setZeroes(matrix: Array<IntArray>) {
        val array = BooleanArray(matrix.size + matrix[0].size) { false }
        for (i in 0 until matrix.size) {
            for (j in 0 until matrix[i].size) {
                array[i] = array[i] || matrix[i][j] == 0
                array[j + matrix.size] = array[j + matrix.size] || matrix[i][j] == 0
            }
        }
        for (i in 0 until matrix.size) {
            if (array[i]) {
                for (j in 0 until matrix[i].size) {
                    matrix[i][j] = 0
                }
            }
        }
        for (i in matrix.size until array.size) {
            if (array[i]) {
                for (j in 0 until matrix.size) {
                    matrix[j][i - matrix.size] = 0
                }
            }
        }
    }
}